File: python/realtimex_sdk/activities.py
"""
Activities Module - HTTP Proxy to RealtimeX Main App
No direct Supabase access - all operations go through Main App
"""

from typing import Any, Dict, List, Optional
import httpx
import os


from .api import PermissionDeniedError, PermissionRequiredError


class ActivitiesModule:
    """CRUD operations for activities via RealtimeX Main App proxy."""

    def __init__(self, realtimex_url: str, app_id: str, app_name: str = None, api_key: str = None):
        self.base_url = realtimex_url.rstrip("/")
        self.app_id = app_id
        self.app_name = app_name or os.environ.get("RTX_APP_NAME", "Local App")
        self.api_key = api_key

    def _get_headers(self) -> Dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if self.app_id:
            headers["x-app-id"] = self.app_id
        return headers

    async def _request_permission(self, permission: str) -> bool:
        """Request a single permission from Electron via internal API."""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.base_url}/api/local-apps/request-permission",
                    json={
                        "app_id": self.app_id,
                        "app_name": self.app_name,
                        "permission": permission,
                    },
                    timeout=60.0  # Long timeout for user interaction
                )
                data = response.json()
                return data.get("granted", False)
        except Exception as e:
            print(f"[SDK] Permission request failed: {e}")
            return False

    async def _handle_response(self, response: httpx.Response, retry_fn) -> Dict[str, Any]:
        """Handle response with permission error handling."""
        if response.status_code == 403:
            data = response.json()
            error_code = data.get("error")
            permission = data.get("permission")
            message = data.get("message")

            if error_code == "PERMISSION_REQUIRED" and permission:
                # Try to get permission from user
                granted = await self._request_permission(permission)

                if granted:
                    # Retry the original request
                    return await retry_fn()
                else:
                    raise PermissionDeniedError(permission, message)

            if error_code == "PERMISSION_DENIED":
                raise PermissionDeniedError(permission, message)

        response.raise_for_status()
        return response.json()

    async def insert(self, raw_data: Dict[str, Any]) -> Dict[str, Any]:
        """Insert a new activity."""
        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.base_url}/activities",
                    headers=self._get_headers(),
                    json={"raw_data": raw_data},
                )
                data = await self._handle_response(response, do_request)
                return data.get("data", data)
        return await do_request()

    async def update(self, id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update an existing activity."""
        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.patch(
                    f"{self.base_url}/activities/{id}",
                    headers=self._get_headers(),
                    json=updates,
                )
                data = await self._handle_response(response, do_request)
                return data.get("data", data)
        return await do_request()

    async def delete(self, id: str) -> None:
        """Delete an activity."""
        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.delete(
                    f"{self.base_url}/activities/{id}",
                    headers=self._get_headers(),
                )
                await self._handle_response(response, do_request)
        await do_request()

    async def get(self, id: str) -> Optional[Dict[str, Any]]:
        """Get an activity by ID."""
        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/activities/{id}",
                    headers=self._get_headers(),
                )
                if response.status_code == 404:
                    return None
                data = await self._handle_response(response, do_request)
                return data.get("data", data)
        return await do_request()

    async def list(
        self,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """List activities with optional filters."""
        params = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status

        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/activities",
                    headers=self._get_headers(),
                    params=params,
                )
                data = await self._handle_response(response, do_request)
                return data.get("data", [])
        return await do_request()

File: typescript/src/modules/activities.ts
/**
 * Activities Module - HTTP Proxy to RealtimeX Main App
 * No direct Supabase access - all operations go through Main App
 */

import { Activity } from '../types';
import { PermissionDeniedError } from './api';

export class ActivitiesModule {
    private baseUrl: string;
    private appId: string;
    private appName: string;
    private apiKey?: string;

    constructor(realtimexUrl: string, appId: string, appName?: string, apiKey?: string) {
        this.baseUrl = realtimexUrl.replace(/\/$/, '');
        this.appId = appId;
        this.appName = appName || process.env.RTX_APP_NAME || 'Local App';
        this.apiKey = apiKey;
    }

    /**
     * Request a single permission from Electron via internal API
     */
    private async requestPermission(permission: string): Promise<boolean> {
        try {
            const response = await fetch(`${this.baseUrl}/api/local-apps/request-permission`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    app_id: this.appId,
                    app_name: this.appName,
                    permission,
                }),
            });
            const data = await response.json();
            return data.granted === true;
        } catch (error) {
            console.error('[SDK] Permission request failed:', error);
            return false;
        }
    }

    private async request<T>(
        path: string,
        options: RequestInit = {}
    ): Promise<T> {
        const url = `${this.baseUrl}${path}`;
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
        };

        if (this.apiKey) {
            headers['Authorization'] = `Bearer ${this.apiKey}`;
        }
        if (this.appId) {
            // Production mode: use x-app-id
            headers['x-app-id'] = this.appId;
        }

        const response = await fetch(url, {
            ...options,
            headers: {
                ...headers,
                ...options.headers,
            },
        });

        const data = await response.json();

        // Handle permission errors
        if (response.status === 403) {
            const errorCode = data.error;
            const permission = data.permission;
            const message = data.message;

            if (errorCode === 'PERMISSION_REQUIRED' && permission) {
                // Try to get permission from user
                const granted = await this.requestPermission(permission);

                if (granted) {
                    // Retry the original request
                    return this.request<T>(path, options);
                } else {
                    throw new PermissionDeniedError(permission, message);
                }
            }

            if (errorCode === 'PERMISSION_DENIED') {
                throw new PermissionDeniedError(permission, message);
            }
        }

        if (!response.ok) {
            throw new Error(data.error || `Request failed: ${response.status}`);
        }
        return data;
    }

    /**
     * Insert a new activity
     */
    async insert(rawData: Record<string, unknown>): Promise<Activity> {
        const result = await this.request<{ data: Activity }>('/activities', {
            method: 'POST',
            body: JSON.stringify({ raw_data: rawData }),
        });
        return result.data;
    }

    /**
     * Update an existing activity
     */
    async update(id: string, updates: Partial<Activity>): Promise<Activity> {
        const result = await this.request<{ data: Activity }>(`/activities/${id}`, {
            method: 'PATCH',
            body: JSON.stringify(updates),
        });
        return result.data;
    }

    /**
     * Delete an activity
     */
    async delete(id: string): Promise<void> {
        await this.request<{ success: boolean }>(`/activities/${id}`, {
            method: 'DELETE',
        });
    }

    /**
     * Get a single activity by ID
     */
    async get(id: string): Promise<Activity | null> {
        try {
            const result = await this.request<{ data: Activity }>(`/activities/${id}`);
            return result.data;
        } catch (error: any) {
            if (error.message?.includes('not found')) return null;
            throw error;
        }
    }

    /**
     * List activities with optional filters
     */
    async list(options?: { status?: string; limit?: number; offset?: number }): Promise<Activity[]> {
        const params = new URLSearchParams();
        if (options?.status) params.set('status', options.status);
        if (options?.limit) params.set('limit', String(options.limit));
        if (options?.offset) params.set('offset', String(options.offset));

        const query = params.toString() ? `?${params}` : '';
        const result = await this.request<{ data: Activity[] }>(`/activities${query}`);
        return result.data;
    }
}
